### PR TITLE
Refactoring project structure

### PR DESCRIPTION
**Pull request by gkyslenko**
Added web elements initialization in constructor for Moderator, Waiter and headers

### CODE REVIEW

File: src/test/java/pages/moderator/ModeratorOwnersPage.java [hunk 1]
```diff
@@ -8,15 +11,22 @@
...
     private String ownerEmail;
 
+    private WebDriver driver = null;
+
     public void setOwnerEmail(String ownerEmail)    {
         this.ownerEmail = ownerEmail;
     }
 
-    private final HeaderGeneralPageComponent headerGlobal = new HeaderGeneralPageComponent();
+    public ModeratorOwnersPage(WebDriver driver)    {
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/moderator/ModeratorOwnersPage.java:9]: There are no elements to initiate

File: src/test/java/pages/waiter/WaiterMainPage.java [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
+    protected WebDriver driver;
 
     public void setOrderNumber(String orderNumber)    {
         this.orderNumber = orderNumber;
     }
 
-    private final HeaderGeneralPageComponent headerGeneralPageComponent = new HeaderGeneralPageComponent();
-    private final HeaderPageComponent waiterHeader = new HeaderPageComponent();
-    private final WaiterOrdersPageComponent orders = new WaiterOrdersPageComponent(orderNumber);
+    public WaiterMainPage(WebDriver driver) {
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/waiter/WaiterMainPage.java:16]: There are no elements to initiate

File: src/test/java/pages/moderator/ModeratorUsersPage.java [hunk 3]
```diff
@@ -8,15 +11,22 @@
...
     private String userEmail;
 
+    private WebDriver driver = null;
+
     public void setUserEmail(String userEmail)    {
         this.userEmail = userEmail;
     }
 
-    private final HeaderGeneralPageComponent headerGlobal = new HeaderGeneralPageComponent();
+    public ModeratorUsersPage (WebDriver driver)    {
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/moderator/ModeratorUsersPage.java:10]: There are no elements to initiate

File: src/test/java/pageComponents/HeaderGeneralPageComponent.java [hunk 4]
```diff
@@ -4,13 +4,22 @@
...
+import org.openqa.selenium.support.PageFactory;
 import pages.HomePage;
 import pages.RestaurantsPage;
 import pages.SignInPage;
 import pages.SignUpPage;
 import pages.waiter.WaiterMainPage;
 
 public class HeaderGeneralPageComponent {
+
+    private WebDriver driver = null;
```
Comment (oleksandranikitiuk) [inline src/test/java/pageComponents/HeaderGeneralPageComponent.java]: Why driver = null?

File: src/test/java/pageComponents/moderator/ModeratorOwnersTablePageComponent.java [hunk 5]
```diff
@@ -1,18 +1,25 @@
...
+
     @FindBy(xpath = "//table/tbody/tr")
     private List<WebElement> owners;
 
     private WebElement neededOwner;
-    public ModeratorOwnersTablePageComponent(String ownerEmail)   {
+    public ModeratorOwnersTablePageComponent(WebDriver driver, String ownerEmail)   {
+        PageFactory.initElements(driver, this);
+        this.driver = driver;
         for(WebElement owner : owners)    {
```
Comment (oleksandranikitiuk) [inline src/test/java/pageComponents/moderator/ModeratorOwnersTablePageComponent.java:26]: You should crate a separate method for this

File: src/test/java/pageComponents/moderator/ModeratorRestaurantsPageComponent.java [hunk 6]
```diff
@@ -1,19 +1,24 @@
...
+    private WebDriver driver;
 
     @FindBy(xpath = "//main/div/div")
     private List<WebElement> restaurants;
 
-    public ModeratorRestaurantsPageComponent(String desiredRestaurantName) {
+    public ModeratorRestaurantsPageComponent(WebDriver driver, String desiredRestaurantName) {
+        PageFactory.initElements(driver, this);
+        this.driver = driver;
         for(WebElement restaurant : restaurants)    {
```
Comment (oleksandranikitiuk) [inline src/test/java/pageComponents/moderator/ModeratorRestaurantsPageComponent.java:23]: Same as comment above

File: src/test/java/pageComponents/moderator/ModeratorUsersTablePageComponent.java [hunk 7]
```diff
@@ -1,18 +1,25 @@
...
+
+    private WebDriver driver;
     @FindBy(xpath = "//table/tbody/tr")
     private List<WebElement> users;
 
-    public ModeratorUsersTablePageComponent(String userEmail)   {
+    public ModeratorUsersTablePageComponent(WebDriver driver, String userEmail)   {
+        PageFactory.initElements(driver, this);
+        this.driver = driver;
         for(WebElement user : users)    {
```
Comment (oleksandranikitiuk) [inline src/test/java/pageComponents/moderator/ModeratorUsersTablePageComponent.java:28]: Same as comment above

File: src/test/java/pageComponents/waiter/WaiterOrdersPageComponent.java [hunk 8]
```diff
@@ -1,19 +1,24 @@
...
 
     private final WaiterOrderInfoPageComponent orderInfo = expandOrder();
     @FindBy(xpath = "//main/div/div")
     private List<WebElement> orders;
 
-    public WaiterOrdersPageComponent(String orderNumber)  {
+    public WaiterOrdersPageComponent(WebDriver driver, String orderNumber)  {
+        PageFactory.initElements(driver, this);
+        this.driver = driver;
         for(WebElement order : orders)  {
```
Comment (oleksandranikitiuk) [inline src/test/java/pageComponents/waiter/WaiterOrdersPageComponent.java:29]: Same as comment above

File: src/test/java/pages/moderator/ModeratorRestaurantsPage.java [hunk 9]
```diff
@@ -10,20 +12,27 @@
 public class ModeratorRestaurantsPage {
     private String restaurantName;
 
+    private WebDriver driver = null;
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/moderator/ModeratorRestaurantsPage.java]: Why driver = null?

File: src/test/java/pages/moderator/ModeratorUsersPage.java [hunk 10]
```diff
@@ -8,15 +11,22 @@
 public class ModeratorUsersPage {
     private String userEmail;
 
+    private WebDriver driver = null;
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/moderator/ModeratorUsersPage.java]: Same as above

**Comment by oleksandranikitiuk** [comment]
Please add constructor to SignIn Page and Home Page

File: src/test/java/pages/moderator/ModeratorOwnersPage.java (hunk 1 above)
Comment (gkyslenko) [reply src/test/java/pages/moderator/ModeratorOwnersPage.java:9]: done

File: src/test/java/pages/waiter/WaiterMainPage.java (hunk 2 above)
Comment (gkyslenko) [reply src/test/java/pages/waiter/WaiterMainPage.java:16]: done

File: src/test/java/pages/moderator/ModeratorUsersPage.java (hunk 3 above)
Comment (gkyslenko) [reply src/test/java/pages/moderator/ModeratorUsersPage.java:10]: done

File: src/test/java/pageComponents/HeaderGeneralPageComponent.java (hunk 4 above)
Comment (gkyslenko) [reply src/test/java/pageComponents/HeaderGeneralPageComponent.java]: done

File: src/test/java/pageComponents/moderator/ModeratorOwnersTablePageComponent.java (hunk 5 above)
Comment (gkyslenko) [reply src/test/java/pageComponents/moderator/ModeratorOwnersTablePageComponent.java:26]: done

File: src/test/java/pageComponents/moderator/ModeratorRestaurantsPageComponent.java (hunk 6 above)
Comment (gkyslenko) [reply src/test/java/pageComponents/moderator/ModeratorRestaurantsPageComponent.java:23]: done

File: src/test/java/pageComponents/moderator/ModeratorUsersTablePageComponent.java (hunk 7 above)
Comment (gkyslenko) [reply src/test/java/pageComponents/moderator/ModeratorUsersTablePageComponent.java:28]: done

File: src/test/java/pageComponents/waiter/WaiterOrdersPageComponent.java (hunk 8 above)
Comment (gkyslenko) [reply src/test/java/pageComponents/waiter/WaiterOrdersPageComponent.java:29]: done

File: src/test/java/pages/moderator/ModeratorRestaurantsPage.java (hunk 9 above)
Comment (gkyslenko) [reply src/test/java/pages/moderator/ModeratorRestaurantsPage.java]: done

File: src/test/java/pages/moderator/ModeratorUsersPage.java (hunk 10 above)
Comment (gkyslenko) [reply src/test/java/pages/moderator/ModeratorUsersPage.java]: done

File: src/test/java/pages/SignInPage.java [hunk 11]
```diff
@@ -6,6 +6,9 @@
 import org.openqa.selenium.support.PageFactory;
 
 public class SignInPage {
+    public SignInPage(WebDriver driver) {
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/SignInPage.java:9]: Please add `private final WebDriver driver` variable

File: src/test/java/pages/HomePage.java [hunk 12]
```diff
@@ -6,6 +6,9 @@
 import org.openqa.selenium.support.PageFactory;
 
 public class HomePage {
+    public HomePage(WebDriver driver)   {
```
Comment (oleksandranikitiuk) [inline src/test/java/pages/HomePage.java:9]: Please add `private final WebDriver driver` variable

**Review by derkachOksana — COMMENTED** [review]
all okay

**Review by derkachOksana — DISMISSED** [review]

File: src/test/java/pages/SignInPage.java (hunk 11 above)
Comment (gkyslenko) [reply src/test/java/pages/SignInPage.java:9]: we aren't using it at this class, do we actually need it?

File: src/test/java/pages/HomePage.java (hunk 12 above)
Comment (gkyslenko) [reply src/test/java/pages/HomePage.java:9]: same as above

Comment (oleksandranikitiuk) [reply src/test/java/pages/HomePage.java:9]: We need this to initiate elements

Comment (gkyslenko) [reply src/test/java/pages/HomePage.java:9]: elements initiation is inside constructor using local variable, that we get as argument to constructor

Comment (oleksandranikitiuk) [reply src/test/java/pages/HomePage.java:9]: Ok, I agree

File: src/test/java/pages/moderator/ModeratorBasePage.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+    public ModeratorBasePage (WebDriver driver)    {
+        PageFactory.initElements(driver, this);
+        this.driver = driver;
+        headerGlobal = new HeaderGeneralPageComponent(driver);
+        header = new ModeratorHeaderPageComponent(driver);
+    }
+
+    public SignInPage logOut()  {
+        return headerGlobal.logOut();
```
Comment (dmatsev) [inline src/test/java/pages/moderator/ModeratorBasePage.java]: You do not need to wrap it with a page object, Page Component class could be public, so you can easily call the component directly:

`page.component.someAction();`

Comment (dmatsev) [reply src/test/java/pages/moderator/ModeratorBasePage.java]: I suppose you have to change it everywhere, where the page component is private or protected.

**Review by dmatsev — CHANGES_REQUESTED** [review]
Please make page components public.

**Review by derkachOksana — APPROVED** [review]

**Review by dmatsev — APPROVED** [review]

**Review by oleksandranikitiuk — APPROVED** [review]

**Pull request merged**